src/lib.rs

```diff
@@ -14,64 +16,71 @@ pub struct CInfo {
-        Ok(s) => {
-            let res = detect(s);
-            match res {
-                Some(info) => {
-                    cinfo.lang = info.lang();
-                    cinfo.script = info.script();
-                    cinfo.confidence = info.confidence();
-                    return 0;
-                },
-                None => {
-                    // Could not detect language
-                    return 1;
-                }
-            }
-        },
-        Err(_) => {
-            // Bad string pointer
-            return 2;
-        }
-    }
+#[no_mangle]
+pub extern "C" fn whatlang_detect(ptr: *const c_char, cinfo: &mut CInfo) -> u8 {
+    let cs = unsafe { CStr::from_ptr(ptr) };
+	detect_internal(cs.to_bytes(), cinfo)
 }
 
 #[no_mangle]
-pub extern fn whatlang_lang_eng_name(lang: Lang, buffer_ptr: *mut c_char) {
-    // Here unwrap is always safe, because whatlang always returns a valid str
-    let s = CString::new(lang.eng_name()).unwrap();
-    unsafe {
-        libc::strcpy(buffer_ptr, s.as_ptr());
-    }
+pub extern "C" fn whatlang_lang_eng_name(lang: Lang, buffer_ptr: *mut c_char) -> size_t {
+	copy_cstr(lang.eng_name(), buffer_ptr)
 }
 
 #[no_mangle]
-pub extern fn whatlang_lang_code(lang: Lang, buffer_ptr: *mut c_char) {
-    // Here unwrap is supposed to be safe, because whatlang return valid str
-    let s = CString::new(lang.code()).unwrap();
-    unsafe {
-        libc::strcpy(buffer_ptr, s.as_ptr());
-    }
+pub extern "C" fn whatlang_lang_code(lang: Lang, buffer_ptr: *mut c_char) -> size_t {
+	copy_cstr(lang.code(), buffer_ptr)
 }
 
 #[no_mangle]
-pub extern fn whatlang_lang_name(lang: Lang, buffer_ptr: *mut c_char) {
-    // Here unwrap is supposed to be safe, because whatlang return valid str
-    let s = CString::new(lang.name()).unwrap();
-    unsafe {
-        libc::strcpy(buffer_ptr, s.as_ptr());
-    }
+pub extern "C" fn whatlang_lang_name(lang: Lang, buffer_ptr: *mut c_char) -> size_t {
+	copy_cstr(lang.name(), buffer_ptr)
 }
 
 #[no_mangle]
-pub extern fn whatlang_script_name(script: Script, buffer_ptr: *mut c_char) {
-    // Here unwrap is supposed to be safe, because whatlang return valid str
-    let s = CString::new(script.name()).unwrap();
-    unsafe {
-        libc::strcpy(buffer_ptr, s.as_ptr());
+pub extern "C" fn whatlang_script_name(script: Script, buffer_ptr: *mut c_char) -> size_t {
+	copy_cstr(script.name(), buffer_ptr)
+}
+
+fn detect_internal(text: &[u8], cinfo: &mut CInfo) -> u8 {
+	match std::str::from_utf8(text) {
+        Ok(s) => {
+			let res = detect(s);
+			match res {
+				Some(info) => {
+					cinfo.lang = info.lang();
+					cinfo.script = info.script();
+					cinfo.confidence = info.confidence();
+					return 0;
+				},
+				None => {
+					// Could not detect language
+					return 1;
+				}
+			}
+        },
+        Err(_) => {
+            // Bad string pointer
+            return 2;
+        }
     }
 }
+
+fn copy_cstr(src: &str, dst: *mut c_char) -> size_t {
```

Comment:
Can we make this function unsafe, as well as all the extern "C" fns?

Reply:
Unsafe on all the extern functions. Got rid of the unneeded unsafe blocks.